### ISSUE TITLE
shuffling order of batches vs order of samples within batch

### ISSUE THREAD

**Issue by nullmightybofo** [cpra/dlvc2018#5]
So, while re-checking some of the stuff of assignment 1, 
I ask myself whether shuffling should be either done such that:
-) samples should be wrapped into a batch sequentially (sample 0, 1, ...) but sequence start should be random for each batch; i.e. batch 1: (samples 0, 1, 2, ..., 199) , batch 2: (samples 1000, 1001, ..., 1199) for batch size 200.
-) samples should be shuffled within each batch , but samples were taken sequentially from the data-set, i.e. batch 1 has shuffling of (samples 0, 1, ..., 199), batch 2 has shuffling of (samples 200, 201, ..., 399), and so on.
-) Combination of both, i.e. randomly sampling from all samples (without replacement) and assigning appropriate number of samples to each batch.

It seems that the third option would presumably destroy any structure/ordering within the data the most.

If I am not mistaken, all options satisfy this claim (when not setting a seed, that is):
"The first sample of the first training batch returned with shuffling must always the different"
from the instructions:
https://github.com/cpra/dlvc2018/blob/master/assignments/assignment1/instructions.md

Opinions, experiences?

**Comment by cpra** [comment]
Option 3 is indeed the best especially with batch normalization. The two others you mention don't cause enough randomization if the dataset is sorted by label - in this case minibatches would likely contain samples of only one class. I mentioned this in the lecture but did not deduct points because the implementation was not clearly stated in the assignment.
